### PR TITLE
[FLINK-23149][table-code-splitter] Introduce the new Java code splitter module

### PR DESCRIPTION
**Pull request by TsReaper**
## What is the purpose of the change

This is the first step of FLINK-23007. This parent issue aims to solve the issue in one shot, that some methods of the generated Java code may grow beyond 64KB and reaches the limit of JVM.

In this first step we introduce a Java code splitter for the generated Java code. This splitter will come into effect after the original Java code is generated, hoping to solve the 64KB problem in one shot. Currently this splitter is not used and we'll activate it in the second step of the parent issue.

In this new module we use antlr to parse Java code. The Java antlr grammar files are copied from the official antlr repository https://github.com/antlr/grammars-v4/tree/master/java/java . 

## Brief change log

 - Introduce the new Java code splitter module

## Verifying this change

This change added tests and can be verified by running the added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **yes**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by TsReaper** [comment]
@JingsongLi please help for the review, thanks.

File: flink-table/flink-table-code-splitter/src/main/antlr4/JavaLexer.g4 [hunk 1]
```diff
@@ -0,0 +1,184 @@
+/**
```
Comment (Airblader) [inline flink-table/flink-table-code-splitter/src/main/antlr4/JavaLexer.g4:1]: The original BSD license has been removed. I think this violates the licensing terms.

File: flink-table/flink-table-code-splitter/pom.xml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		<version>1.14-SNAPSHOT</version>
+		<relativePath>..</relativePath>
+	</parent>
+
+	<artifactId>flink-table-code-splitter</artifactId>
+	<name>Flink : Table : Code Splitter </name>
+	<description>
+		This module contains a tool to split generated Java code
+		so that each method does not exceed the limit of 64KB.
+		Antlr grammar files are copied from the official antlr repository
```
Comment (Airblader) [inline flink-table/flink-table-code-splitter/pom.xml:36]: Given that files are copied verbatim from other projects, how will this be handled in terms of licensing?

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/CodeSplitUtil.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+package org.apache.flink.table.codesplit;
+
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.misc.Interval;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/** Utils for rewriters. */
+public class CodeSplitUtil {
```
Comment (Airblader) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/CodeSplitUtil.java:27]: All classes seem to be missing `@Internal` annotations.

File: flink-table/flink-table-code-splitter/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-table/flink-table-code-splitter/pom.xml:36]: Antlr has the BSD 3-clause license. https://github.com/antlr/antlr4/blob/master/LICENSE.txt
That's a category A license: https://www.apache.org/legal/resolved.html#category-a, which allows us to include code into our project.

I would recommend the following:
**a) Mark the copied code**
Add a comment like this if an entire file got copied: https://github.com/apache/flink/blob/99c2a415e9eeefafacf70762b6f54070f7911ceb/flink-filesystems/flink-fs-hadoop-shaded/src/main/java/org/apache/hadoop/util/NativeCodeLoader.java#L21-L28
If only parts of a file got copied, or if it is a file with mixed origin (parts are self-written/from Flink, the rest is copied), I would clearly indicate where the copied part starts and ends. It is important that people can see which code is from us, and which code is copied.
**b) Add to NOTICE file of the module**
Add an entry to the NOTICE file of the affected module, like this: https://github.com/apache/flink/blob/master/NOTICE#L17-L22 or 
**c) Ship the ANTLR license file in the module**
File: https://github.com/antlr/antlr4/blob/master/LICENSE.txt

[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-code-splitter/pom.xml]: Thanks @rmetzger !

**Comment by TsReaper** [comment]
@flinkbot run azure

File: flink-table/flink-table-code-splitter/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-code-splitter/pom.xml]: @TsReaper You need copy ANTLR license to the code-splitter module too. A example is:
https://github.com/apache/flink/tree/master/flink-connectors/flink-sql-connector-hive-1.2.2/src/main/resources/META-INF/licenses

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/DeclarationRewriter.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.Stack;
+
+/** Extract and rename local variables to member variables. */
+@Internal
+public class DeclarationRewriter {
```
Comment (JingsongLi) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/DeclarationRewriter.java:38]: It is better to add a rewrite example in the comment. Other classes also need.

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/JavaCodeSplitter.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                    "JavaCodeSplitter failed. This is a bug. Please file an issue.", t);
+        }
+    }
+
+    private static String splitImpl(String code, int maxMethodLength, int maxClassMemberCount) {
+        checkArgument(code != null && !code.isEmpty(), "code cannot be empty");
+        checkArgument(maxMethodLength > 0);
+        checkArgument(maxClassMemberCount > 0);
+
+        Optional<String> declarationCode = new DeclarationRewriter(code, maxMethodLength).rewrite();
```
Comment (JingsongLi) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/JavaCodeSplitter.java:48]: Use lambda:
```
        return new DeclarationRewriter(code, maxMethodLength)
                .rewrite()
                .map(text -> new IfStatementRewriter(text, maxMethodLength).rewrite())
                .map(text -> new FunctionSplitter(text, maxMethodLength).rewrite())
                .map(text -> new MemberFieldRewriter(text, maxClassMemberCount).rewrite())
                .orElse(code);
```

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

**Pull request merged**